// ngx-joyride/esm5/lib/components/arrow/arrow.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/arrow/arrow.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from "@angular/core";
var JoyrideArrowComponent = /** @class */ (function () {
    function JoyrideArrowComponent() {
        this.position = 'top';
    }
    JoyrideArrowComponent.decorators = [
        { type: Component, args: [{
                    selector: 'joyride-arrow',
                    template: "<div [class.joyride-arrow__top]=\"position == 'top'\"\r\n     [class.joyride-arrow__bottom]=\"position == 'bottom'\"\r\n     [class.joyride-arrow__left]=\"position == 'left'\"\r\n     [class.joyride-arrow__right]=\"position == 'right'\">\r\n</div>",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".joyride-arrow__top{border-left:11px solid transparent;border-right:11px solid transparent;border-bottom:11px solid #fff}.joyride-arrow__bottom{border-left:11px solid transparent;border-right:11px solid transparent;border-top:11px solid #fff}.joyride-arrow__right{border-left:11px solid #fff;border-bottom:11px solid transparent;border-top:11px solid transparent}.joyride-arrow__left{border-right:11px solid #fff;border-top:11px solid transparent;border-bottom:11px solid transparent}"]
                }] }
    ];
    JoyrideArrowComponent.propDecorators = {
        position: [{ type: Input }]
    };
    return JoyrideArrowComponent;
}());
export { JoyrideArrowComponent };
if (false) {
    /** @type {?} */
    JoyrideArrowComponent.prototype.position;
}
