// ngx-joyride/esm5/lib/services/dom.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/dom.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
var DomRefService = /** @class */ (function () {
    function DomRefService(platformId) {
        this.platformId = platformId;
        this.fakeDocument = (/** @type {?} */ ({ body: {}, documentElement: {} }));
        this.fakeWindow = (/** @type {?} */ ({ document: this.fakeDocument, navigator: {} }));
    }
    /**
     * @return {?}
     */
    DomRefService.prototype.getNativeWindow = /**
     * @return {?}
     */
    function () {
        if (isPlatformBrowser(this.platformId))
            return window;
        else
            return this.fakeWindow;
    };
    /**
     * @return {?}
     */
    DomRefService.prototype.getNativeDocument = /**
     * @return {?}
     */
    function () {
        if (isPlatformBrowser(this.platformId))
            return document;
        else
            return this.fakeDocument;
    };
    DomRefService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DomRefService.ctorParameters = function () { return [
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return DomRefService;
}());
export { DomRefService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DomRefService.prototype.fakeDocument;
    /**
     * @type {?}
     * @private
     */
    DomRefService.prototype.fakeWindow;
    /**
     * @type {?}
     * @private
     */
    DomRefService.prototype.platformId;
}
