// ngx-joyride/esm5/lib/services/event-listener.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/event-listener.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, RendererFactory2 } from '@angular/core';
import { Subject } from 'rxjs';
import { DomRefService } from './dom.service';
var Scroll = /** @class */ (function () {
    function Scroll() {
    }
    return Scroll;
}());
export { Scroll };
if (false) {
    /** @type {?} */
    Scroll.prototype.scrollX;
    /** @type {?} */
    Scroll.prototype.scrollY;
}
var EventListenerService = /** @class */ (function () {
    function EventListenerService(rendererFactory, DOMService) {
        this.rendererFactory = rendererFactory;
        this.DOMService = DOMService;
        this.scrollEvent = new Subject();
        this.resizeEvent = new Subject();
        this.renderer = rendererFactory.createRenderer(null, null);
    }
    /**
     * @return {?}
     */
    EventListenerService.prototype.startListeningScrollEvents = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.scrollUnlisten = this.renderer.listen('document', 'scroll', (/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) {
            _this.scrollEvent.next({
                scrollX: _this.DOMService.getNativeWindow().pageXOffset,
                scrollY: _this.DOMService.getNativeWindow().pageYOffset
            });
        }));
    };
    /**
     * @return {?}
     */
    EventListenerService.prototype.startListeningResizeEvents = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.resizeUnlisten = this.renderer.listen('window', 'resize', (/**
         * @param {?} evt
         * @return {?}
         */
        function (evt) {
            _this.resizeEvent.next(evt);
        }));
    };
    /**
     * @return {?}
     */
    EventListenerService.prototype.stopListeningScrollEvents = /**
     * @return {?}
     */
    function () {
        this.scrollUnlisten();
    };
    /**
     * @return {?}
     */
    EventListenerService.prototype.stopListeningResizeEvents = /**
     * @return {?}
     */
    function () {
        this.resizeUnlisten();
    };
    EventListenerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EventListenerService.ctorParameters = function () { return [
        { type: RendererFactory2 },
        { type: DomRefService }
    ]; };
    return EventListenerService;
}());
export { EventListenerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    EventListenerService.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    EventListenerService.prototype.scrollUnlisten;
    /**
     * @type {?}
     * @private
     */
    EventListenerService.prototype.resizeUnlisten;
    /** @type {?} */
    EventListenerService.prototype.scrollEvent;
    /** @type {?} */
    EventListenerService.prototype.resizeEvent;
    /**
     * @type {?}
     * @private
     */
    EventListenerService.prototype.rendererFactory;
    /**
     * @type {?}
     * @private
     */
    EventListenerService.prototype.DOMService;
}
