// ngx-joyride/esm5/lib/services/joyride-step.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/joyride-step.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { JoyrideBackdropService } from './joyride-backdrop.service';
import { EventListenerService } from './event-listener.service';
import { JoyrideStepsContainerService, StepActionType } from './joyride-steps-container.service';
import { DocumentService } from './document.service';
import { StepDrawerService } from './step-drawer.service';
import { DomRefService } from './dom.service';
import { NO_POSITION } from '../directives/joyride.directive';
import { JoyrideOptionsService } from './joyride-options.service';
import { Router } from '@angular/router';
import { ReplaySubject } from 'rxjs';
import { JoyrideStepDoesNotExist, JoyrideStepOutOfRange } from '../models/joyride-error.class';
import { LoggerService } from './logger.service';
/** @type {?} */
var SCROLLBAR_SIZE = 20;
/** @type {?} */
export var DISTANCE_FROM_TARGET = 15;
/** @type {?} */
export var ARROW_SIZE = 10;
/**
 * @record
 */
export function IJoyrideStepService() { }
if (false) {
    /**
     * @return {?}
     */
    IJoyrideStepService.prototype.startTour = function () { };
    /**
     * @return {?}
     */
    IJoyrideStepService.prototype.close = function () { };
    /**
     * @return {?}
     */
    IJoyrideStepService.prototype.prev = function () { };
    /**
     * @return {?}
     */
    IJoyrideStepService.prototype.next = function () { };
}
var JoyrideStepService = /** @class */ (function () {
    function JoyrideStepService(backDropService, eventListener, stepsContainerService, documentService, DOMService, stepDrawerService, optionsService, router, logger) {
        this.backDropService = backDropService;
        this.eventListener = eventListener;
        this.stepsContainerService = stepsContainerService;
        this.documentService = documentService;
        this.DOMService = DOMService;
        this.stepDrawerService = stepDrawerService;
        this.optionsService = optionsService;
        this.router = router;
        this.logger = logger;
        this.winTopPosition = 0;
        this.winBottomPosition = 0;
        this.stepsObserver = new ReplaySubject();
        this.initViewportPositions();
        this.subscribeToScrollEvents();
        this.subscribeToResizeEvents();
    }
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.initViewportPositions = /**
     * @private
     * @return {?}
     */
    function () {
        this.winTopPosition = 0;
        this.winBottomPosition = this.DOMService.getNativeWindow().innerHeight - SCROLLBAR_SIZE;
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.subscribeToScrollEvents = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.eventListener.startListeningScrollEvents();
        this.eventListener.scrollEvent.subscribe((/**
         * @param {?} scroll
         * @return {?}
         */
        function (scroll) {
            _this.winTopPosition = scroll.scrollY;
            _this.winBottomPosition = _this.winTopPosition + _this.DOMService.getNativeWindow().innerHeight - SCROLLBAR_SIZE;
            if (_this.currentStep)
                _this.backDropService.redraw(_this.currentStep, scroll);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.subscribeToResizeEvents = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.eventListener.resizeEvent.subscribe((/**
         * @return {?}
         */
        function () {
            if (_this.currentStep)
                _this.backDropService.redrawTarget(_this.currentStep);
        }));
    };
    /**
     * @private
     * @param {?} step
     * @return {?}
     */
    JoyrideStepService.prototype.drawStep = /**
     * @private
     * @param {?} step
     * @return {?}
     */
    function (step) {
        step.position = step.position === NO_POSITION ? this.optionsService.getStepDefaultPosition() : step.position;
        this.stepDrawerService.draw(step);
    };
    /**
     * @return {?}
     */
    JoyrideStepService.prototype.startTour = /**
     * @return {?}
     */
    function () {
        this.stepsObserver = new ReplaySubject();
        this.stepsContainerService.init();
        this.documentService.setDocumentHeight();
        this.tryShowStep(StepActionType.NEXT);
        this.eventListener.startListeningResizeEvents();
        this.subscribeToStepsUpdates();
        return this.stepsObserver.asObservable();
    };
    /**
     * @return {?}
     */
    JoyrideStepService.prototype.close = /**
     * @return {?}
     */
    function () {
        this.removeCurrentStep();
        this.notifyTourIsFinished();
        this.DOMService.getNativeWindow().scrollTo(0, 0);
        this.eventListener.stopListeningResizeEvents();
        this.backDropService.remove();
    };
    /**
     * @return {?}
     */
    JoyrideStepService.prototype.prev = /**
     * @return {?}
     */
    function () {
        this.removeCurrentStep();
        this.currentStep.prevCliked.emit();
        this.tryShowStep(StepActionType.PREV);
    };
    /**
     * @return {?}
     */
    JoyrideStepService.prototype.next = /**
     * @return {?}
     */
    function () {
        this.removeCurrentStep();
        this.currentStep.nextClicked.emit();
        this.tryShowStep(StepActionType.NEXT);
    };
    /**
     * @private
     * @param {?} action
     * @return {?}
     */
    JoyrideStepService.prototype.navigateToStepPage = /**
     * @private
     * @param {?} action
     * @return {?}
     */
    function (action) {
        /** @type {?} */
        var stepRoute = this.stepsContainerService.getStepRoute(action);
        if (stepRoute) {
            this.router.navigate([stepRoute]);
        }
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.subscribeToStepsUpdates = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.stepsContainerService.stepHasBeenModified.subscribe((/**
         * @param {?} updatedStep
         * @return {?}
         */
        function (updatedStep) {
            if (_this.currentStep && _this.currentStep.name === updatedStep.name) {
                _this.currentStep = updatedStep;
            }
        }));
    };
    /**
     * @private
     * @param {?} actionType
     * @return {?}
     */
    JoyrideStepService.prototype.tryShowStep = /**
     * @private
     * @param {?} actionType
     * @return {?}
     */
    function (actionType) {
        var _this = this;
        this.navigateToStepPage(actionType);
        /** @type {?} */
        var timeout = this.optionsService.getWaitingTime();
        if (timeout > 100)
            this.backDropService.remove();
        setTimeout((/**
         * @return {?}
         */
        function () {
            try {
                _this.showStep(actionType);
            }
            catch (error) {
                if (error instanceof JoyrideStepDoesNotExist) {
                    _this.tryShowStep(actionType);
                }
                else if (error instanceof JoyrideStepOutOfRange) {
                    _this.logger.error('Forcing the tour closure: First or Last step not found in the DOM.');
                    _this.close();
                }
                else {
                    throw new Error(error);
                }
            }
        }), timeout);
    };
    /**
     * @private
     * @param {?} actionType
     * @return {?}
     */
    JoyrideStepService.prototype.showStep = /**
     * @private
     * @param {?} actionType
     * @return {?}
     */
    function (actionType) {
        this.currentStep = this.stepsContainerService.get(actionType);
        if (this.currentStep == null)
            throw new JoyrideStepDoesNotExist('');
        // Scroll the element to get it visible if it's in a scrollable element
        this.scrollIfElementBeyondOtherElements();
        this.backDropService.draw(this.currentStep);
        this.drawStep(this.currentStep);
        this.scrollIfStepAndTargetAreNotVisible();
        this.notifyStepClicked(actionType);
    };
    /**
     * @private
     * @param {?} actionType
     * @return {?}
     */
    JoyrideStepService.prototype.notifyStepClicked = /**
     * @private
     * @param {?} actionType
     * @return {?}
     */
    function (actionType) {
        /** @type {?} */
        var stepInfo = {
            number: this.stepsContainerService.getStepNumber(this.currentStep.name),
            name: this.currentStep.name,
            route: this.currentStep.route,
            actionType: actionType
        };
        this.stepsObserver.next(stepInfo);
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.notifyTourIsFinished = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.currentStep)
            this.currentStep.tourDone.emit();
        this.stepsObserver.complete();
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.removeCurrentStep = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.currentStep)
            this.stepDrawerService.remove(this.currentStep);
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.scrollIfStepAndTargetAreNotVisible = /**
     * @private
     * @return {?}
     */
    function () {
        this.scrollWhenTargetOrStepAreHiddenBottom();
        this.scrollWhenTargetOrStepAreHiddenTop();
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.scrollWhenTargetOrStepAreHiddenBottom = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var totalTargetBottom = this.getMaxTargetAndStepBottomPosition();
        if (totalTargetBottom > this.winBottomPosition) {
            this.DOMService.getNativeWindow().scrollBy(0, totalTargetBottom - this.winBottomPosition);
        }
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.scrollWhenTargetOrStepAreHiddenTop = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var totalTargetTop = this.getMaxTargetAndStepTopPosition();
        if (totalTargetTop < this.winTopPosition) {
            this.DOMService.getNativeWindow().scrollBy(0, totalTargetTop - this.winTopPosition);
        }
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.getMaxTargetAndStepBottomPosition = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var targetAbsoluteTop = this.documentService.getElementAbsoluteTop(this.currentStep.targetViewContainer.element);
        if (this.currentStep.position === 'top') {
            return targetAbsoluteTop + this.currentStep.stepInstance.targetHeight;
        }
        else if (this.currentStep.position === 'bottom') {
            return (targetAbsoluteTop +
                this.currentStep.stepInstance.targetHeight +
                this.currentStep.stepInstance.stepHeight +
                ARROW_SIZE +
                DISTANCE_FROM_TARGET);
        }
        else if (this.currentStep.position === 'right' || this.currentStep.position === 'left') {
            return Math.max(targetAbsoluteTop + this.currentStep.stepInstance.targetHeight, targetAbsoluteTop + this.currentStep.stepInstance.targetHeight / 2 + this.currentStep.stepInstance.stepHeight / 2);
        }
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.getMaxTargetAndStepTopPosition = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var targetAbsoluteTop = this.documentService.getElementAbsoluteTop(this.currentStep.targetViewContainer.element);
        if (this.currentStep.position === 'top') {
            return targetAbsoluteTop - (this.currentStep.stepInstance.stepHeight + ARROW_SIZE + DISTANCE_FROM_TARGET);
        }
        else if (this.currentStep.position === 'bottom') {
            return targetAbsoluteTop;
        }
        else if (this.currentStep.position === 'right' || this.currentStep.position === 'left') {
            return Math.min(targetAbsoluteTop, targetAbsoluteTop + this.currentStep.stepInstance.targetHeight / 2 - this.currentStep.stepInstance.stepHeight / 2);
        }
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.scrollIfElementBeyondOtherElements = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.isElementBeyondOthers() === 2) {
            this.documentService.scrollToTheTop(this.currentStep.targetViewContainer.element);
        }
        if (this.isElementBeyondOthers() === 2) {
            this.documentService.scrollToTheBottom(this.currentStep.targetViewContainer.element);
        }
        if (this.isElementBeyondOthers() === 1 && this.documentService.isParentScrollable(this.currentStep.targetViewContainer.element)) {
            this.documentService.scrollIntoView(this.currentStep.targetViewContainer.element, this.currentStep.isElementOrAncestorFixed);
        }
        if (this.isElementBeyondOthers() === 1 && this.documentService.isParentScrollable(this.currentStep.targetViewContainer.element)) {
            this.currentStep.targetViewContainer.element.nativeElement.scrollIntoView();
        }
    };
    /**
     * @private
     * @return {?}
     */
    JoyrideStepService.prototype.isElementBeyondOthers = /**
     * @private
     * @return {?}
     */
    function () {
        return this.documentService.isElementBeyondOthers(this.currentStep.targetViewContainer.element, this.currentStep.isElementOrAncestorFixed, 'backdrop');
    };
    JoyrideStepService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    JoyrideStepService.ctorParameters = function () { return [
        { type: JoyrideBackdropService },
        { type: EventListenerService },
        { type: JoyrideStepsContainerService },
        { type: DocumentService },
        { type: DomRefService },
        { type: StepDrawerService },
        { type: JoyrideOptionsService },
        { type: Router },
        { type: LoggerService }
    ]; };
    return JoyrideStepService;
}());
export { JoyrideStepService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.currentStep;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.winTopPosition;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.winBottomPosition;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.stepsObserver;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.backDropService;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.eventListener;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.stepsContainerService;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.documentService;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.DOMService;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.stepDrawerService;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.optionsService;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    JoyrideStepService.prototype.logger;
}
