// ngx-joyride/esm5/lib/services/joyride-options.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/joyride-options.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of, Observable } from 'rxjs';
/** @type {?} */
export var DEFAULT_THEME_COLOR = '#3b5560';
/** @type {?} */
export var STEP_DEFAULT_POSITION = 'bottom';
/** @type {?} */
export var DEFAULT_TIMEOUT_BETWEEN_STEPS = 1;
var ObservableCustomTexts = /** @class */ (function () {
    function ObservableCustomTexts() {
    }
    return ObservableCustomTexts;
}());
export { ObservableCustomTexts };
if (false) {
    /** @type {?} */
    ObservableCustomTexts.prototype.prev;
    /** @type {?} */
    ObservableCustomTexts.prototype.next;
    /** @type {?} */
    ObservableCustomTexts.prototype.done;
    /** @type {?} */
    ObservableCustomTexts.prototype.close;
}
/** @type {?} */
export var DEFAULT_TEXTS = {
    prev: of('prev'),
    next: of('next'),
    done: of('done'),
    close: of(null)
};
/**
 * @record
 */
export function IJoyrideOptionsService() { }
if (false) {
    /**
     * @param {?} options
     * @return {?}
     */
    IJoyrideOptionsService.prototype.setOptions = function (options) { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.getBackdropColor = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.getThemeColor = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.getStepDefaultPosition = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.getStepsOrder = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.getFirstStep = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.getWaitingTime = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.areLogsEnabled = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.isCounterVisible = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.isPrevButtonVisible = function () { };
    /**
     * @return {?}
     */
    IJoyrideOptionsService.prototype.getCustomTexts = function () { };
}
var JoyrideOptionsService = /** @class */ (function () {
    function JoyrideOptionsService() {
        this.themeColor = DEFAULT_THEME_COLOR;
        this.stepDefaultPosition = STEP_DEFAULT_POSITION;
        this.logsEnabled = false;
        this.showCounter = true;
        this.showPrevButton = true;
        this.stepsOrder = [];
    }
    /**
     * @param {?} options
     * @return {?}
     */
    JoyrideOptionsService.prototype.setOptions = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.stepsOrder = options.steps;
        this.stepDefaultPosition = options.stepDefaultPosition
            ? options.stepDefaultPosition
            : this.stepDefaultPosition;
        this.logsEnabled =
            typeof options.logsEnabled !== 'undefined'
                ? options.logsEnabled
                : this.logsEnabled;
        this.showCounter =
            typeof options.showCounter !== 'undefined'
                ? options.showCounter
                : this.showCounter;
        this.showPrevButton =
            typeof options.showPrevButton !== 'undefined'
                ? options.showPrevButton
                : this.showPrevButton;
        this.themeColor = options.themeColor
            ? options.themeColor
            : this.themeColor;
        this.firstStep = options.startWith;
        this.waitingTime =
            typeof options.waitingTime !== 'undefined'
                ? options.waitingTime
                : DEFAULT_TIMEOUT_BETWEEN_STEPS;
        typeof options.customTexts !== 'undefined'
            ? this.setCustomText(options.customTexts)
            : this.setCustomText(DEFAULT_TEXTS);
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.getBackdropColor = /**
     * @return {?}
     */
    function () {
        return this.hexToRgb(this.themeColor);
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.getThemeColor = /**
     * @return {?}
     */
    function () {
        return this.themeColor;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.getStepDefaultPosition = /**
     * @return {?}
     */
    function () {
        return this.stepDefaultPosition;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.getStepsOrder = /**
     * @return {?}
     */
    function () {
        return this.stepsOrder;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.getFirstStep = /**
     * @return {?}
     */
    function () {
        return this.firstStep;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.getWaitingTime = /**
     * @return {?}
     */
    function () {
        return this.waitingTime;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.areLogsEnabled = /**
     * @return {?}
     */
    function () {
        return this.logsEnabled;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.isCounterVisible = /**
     * @return {?}
     */
    function () {
        return this.showCounter;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.isPrevButtonVisible = /**
     * @return {?}
     */
    function () {
        return this.showPrevButton;
    };
    /**
     * @return {?}
     */
    JoyrideOptionsService.prototype.getCustomTexts = /**
     * @return {?}
     */
    function () {
        return this.customTexts;
    };
    /**
     * @private
     * @param {?} texts
     * @return {?}
     */
    JoyrideOptionsService.prototype.setCustomText = /**
     * @private
     * @param {?} texts
     * @return {?}
     */
    function (texts) {
        /** @type {?} */
        var prev;
        /** @type {?} */
        var next;
        /** @type {?} */
        var done;
        /** @type {?} */
        var close;
        prev = texts.prev ? texts.prev : DEFAULT_TEXTS.prev;
        next = texts.next ? texts.next : DEFAULT_TEXTS.next;
        done = texts.done ? texts.done : DEFAULT_TEXTS.done;
        close = texts.close ? texts.close : DEFAULT_TEXTS.close;
        this.customTexts = (/** @type {?} */ ({
            prev: this.toObservable(prev),
            next: this.toObservable(next),
            done: this.toObservable(done),
            close: this.toObservable(close)
        }));
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    JoyrideOptionsService.prototype.toObservable = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value instanceof Observable ? value : of(value);
    };
    /**
     * @private
     * @param {?} hex
     * @return {?}
     */
    JoyrideOptionsService.prototype.hexToRgb = /**
     * @private
     * @param {?} hex
     * @return {?}
     */
    function (hex) {
        /** @type {?} */
        var shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
        hex = hex.replace(shorthandRegex, (/**
         * @param {?} m
         * @param {?} r
         * @param {?} g
         * @param {?} b
         * @return {?}
         */
        function (m, r, g, b) {
            return r + r + g + g + b + b;
        }));
        /** @type {?} */
        var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
        return result
            ? parseInt(result[1], 16) + ", " + parseInt(result[2], 16) + ", " + parseInt(result[3], 16)
            : null;
    };
    JoyrideOptionsService.decorators = [
        { type: Injectable }
    ];
    return JoyrideOptionsService;
}());
export { JoyrideOptionsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.themeColor;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.stepDefaultPosition;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.logsEnabled;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.showCounter;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.showPrevButton;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.stepsOrder;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.firstStep;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.waitingTime;
    /**
     * @type {?}
     * @private
     */
    JoyrideOptionsService.prototype.customTexts;
}
