// ngx-joyride/esm5/lib/services/joyride.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/joyride.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { JoyrideStepService } from './joyride-step.service';
import { JoyrideOptionsService } from './joyride-options.service';
import { of } from 'rxjs';
import { finalize } from 'rxjs/operators';
import { JoyrideStepInfo } from '../models/joyride-step-info.class';
import { isPlatformBrowser } from '@angular/common';
var JoyrideService = /** @class */ (function () {
    function JoyrideService(platformId, stepService, optionsService) {
        this.platformId = platformId;
        this.stepService = stepService;
        this.optionsService = optionsService;
        this.tourInProgress = false;
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    JoyrideService.prototype.startTour = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        var _this = this;
        if (!isPlatformBrowser(this.platformId)) {
            return of(new JoyrideStepInfo());
        }
        if (!this.tourInProgress) {
            this.tourInProgress = true;
            if (options) {
                this.optionsService.setOptions(options);
            }
            this.tour$ = this.stepService.startTour().pipe(finalize((/**
             * @return {?}
             */
            function () { return (_this.tourInProgress = false); })));
            this.tour$.subscribe();
        }
        return this.tour$;
    };
    /**
     * @return {?}
     */
    JoyrideService.prototype.closeTour = /**
     * @return {?}
     */
    function () {
        if (this.isTourInProgress())
            this.stepService.close();
    };
    /**
     * @return {?}
     */
    JoyrideService.prototype.isTourInProgress = /**
     * @return {?}
     */
    function () {
        return this.tourInProgress;
    };
    JoyrideService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    JoyrideService.ctorParameters = function () { return [
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: JoyrideStepService },
        { type: JoyrideOptionsService }
    ]; };
    return JoyrideService;
}());
export { JoyrideService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    JoyrideService.prototype.tourInProgress;
    /**
     * @type {?}
     * @private
     */
    JoyrideService.prototype.tour$;
    /**
     * @type {?}
     * @private
     */
    JoyrideService.prototype.platformId;
    /**
     * @type {?}
     * @private
     */
    JoyrideService.prototype.stepService;
    /**
     * @type {?}
     * @private
     */
    JoyrideService.prototype.optionsService;
}
