// ngx-joyride/esm5/lib/services/logger.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/logger.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { JoyrideOptionsService } from './joyride-options.service';
/** @type {?} */
var JOYRIDE = 'ngx-joyride:::';
var LoggerService = /** @class */ (function () {
    function LoggerService(optionService) {
        this.optionService = optionService;
    }
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    LoggerService.prototype.debug = /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    function (message, data) {
        if (data === void 0) { data = ""; }
        if (this.optionService.areLogsEnabled()) {
            console.debug(JOYRIDE + message, data);
        }
    };
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    LoggerService.prototype.info = /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    function (message, data) {
        if (data === void 0) { data = ""; }
        if (this.optionService.areLogsEnabled()) {
            console.info(JOYRIDE + message, data);
        }
    };
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    LoggerService.prototype.warn = /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    function (message, data) {
        if (data === void 0) { data = ""; }
        if (this.optionService.areLogsEnabled()) {
            console.warn(JOYRIDE + message, data);
        }
    };
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    LoggerService.prototype.error = /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    function (message, data) {
        if (data === void 0) { data = ""; }
        if (this.optionService.areLogsEnabled()) {
            console.error(JOYRIDE + message, data);
        }
    };
    LoggerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LoggerService.ctorParameters = function () { return [
        { type: JoyrideOptionsService }
    ]; };
    return LoggerService;
}());
export { LoggerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LoggerService.prototype.optionService;
}
