// ngx-joyride/esm5/lib/services/templates.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/templates.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var TemplatesService = /** @class */ (function () {
    function TemplatesService() {
    }
    /**
     * @param {?} template
     * @return {?}
     */
    TemplatesService.prototype.setPrevButton = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        this._prevButton = template;
    };
    /**
     * @return {?}
     */
    TemplatesService.prototype.getPrevButton = /**
     * @return {?}
     */
    function () {
        return this._prevButton;
    };
    /**
     * @param {?} template
     * @return {?}
     */
    TemplatesService.prototype.setNextButton = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        this._nextButton = template;
    };
    /**
     * @return {?}
     */
    TemplatesService.prototype.getNextButton = /**
     * @return {?}
     */
    function () {
        return this._nextButton;
    };
    /**
     * @param {?} template
     * @return {?}
     */
    TemplatesService.prototype.setDoneButton = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        this._doneButton = template;
    };
    /**
     * @return {?}
     */
    TemplatesService.prototype.getDoneButton = /**
     * @return {?}
     */
    function () {
        return this._doneButton;
    };
    /**
     * @param {?} template
     * @return {?}
     */
    TemplatesService.prototype.setCounter = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        this._counter = template;
    };
    /**
     * @return {?}
     */
    TemplatesService.prototype.getCounter = /**
     * @return {?}
     */
    function () {
        return this._counter;
    };
    TemplatesService.decorators = [
        { type: Injectable }
    ];
    return TemplatesService;
}());
export { TemplatesService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._prevButton;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._nextButton;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._doneButton;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._counter;
}
