// ngx-joyride/esm2015/lib/services/logger.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/logger.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { JoyrideOptionsService } from './joyride-options.service';
/** @type {?} */
const JOYRIDE = 'ngx-joyride:::';
export class LoggerService {
    /**
     * @param {?} optionService
     */
    constructor(optionService) {
        this.optionService = optionService;
    }
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    debug(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.debug(JOYRIDE + message, data);
        }
    }
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    info(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.info(JOYRIDE + message, data);
        }
    }
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    warn(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.warn(JOYRIDE + message, data);
        }
    }
    /**
     * @param {?=} message
     * @param {?=} data
     * @return {?}
     */
    error(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.error(JOYRIDE + message, data);
        }
    }
}
LoggerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LoggerService.ctorParameters = () => [
    { type: JoyrideOptionsService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    LoggerService.prototype.optionService;
}
