// ngx-joyride/esm2015/lib/services/step-drawer.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/step-drawer.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, ApplicationRef, Injector } from '@angular/core';
import { JoyrideStepComponent } from '../components';
export class StepDrawerService {
    /**
     * @param {?} componentFactoryResolver
     * @param {?} appRef
     * @param {?} injector
     */
    constructor(componentFactoryResolver, appRef, injector) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.injector = injector;
        this.refMap = {};
    }
    /**
     * @param {?} step
     * @return {?}
     */
    draw(step) {
        // 1. Create a component reference from the component
        /** @type {?} */
        const ref = this.componentFactoryResolver
            .resolveComponentFactory(JoyrideStepComponent)
            .create(this.injector);
        // 2. Attach component to the appRef so that it's inside the ng component tree
        this.appRef.attachView(ref.hostView);
        // 3. Get DOM element from component
        /** @type {?} */
        const domElem = (/** @type {?} */ (((/** @type {?} */ (ref.hostView)))
            .rootNodes[0]));
        // 4. Append DOM element to the body
        document.body.appendChild(domElem);
        /** @type {?} */
        const instance = ref.instance;
        instance.step = step;
        ref.changeDetectorRef.detectChanges();
        step.stepInstance = instance;
        this.refMap[step.name] = ref;
    }
    /**
     * @param {?} step
     * @return {?}
     */
    remove(step) {
        this.appRef.detachView(this.refMap[step.name].hostView);
        this.refMap[step.name].destroy();
    }
}
StepDrawerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
StepDrawerService.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: ApplicationRef },
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    StepDrawerService.prototype.refMap;
    /**
     * @type {?}
     * @private
     */
    StepDrawerService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    StepDrawerService.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    StepDrawerService.prototype.injector;
}
