// ngx-joyride/esm2015/lib/services/templates.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/templates.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class TemplatesService {
    /**
     * @param {?} template
     * @return {?}
     */
    setPrevButton(template) {
        this._prevButton = template;
    }
    /**
     * @return {?}
     */
    getPrevButton() {
        return this._prevButton;
    }
    /**
     * @param {?} template
     * @return {?}
     */
    setNextButton(template) {
        this._nextButton = template;
    }
    /**
     * @return {?}
     */
    getNextButton() {
        return this._nextButton;
    }
    /**
     * @param {?} template
     * @return {?}
     */
    setDoneButton(template) {
        this._doneButton = template;
    }
    /**
     * @return {?}
     */
    getDoneButton() {
        return this._doneButton;
    }
    /**
     * @param {?} template
     * @return {?}
     */
    setCounter(template) {
        this._counter = template;
    }
    /**
     * @return {?}
     */
    getCounter() {
        return this._counter;
    }
}
TemplatesService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._prevButton;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._nextButton;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._doneButton;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._counter;
}
